man/systemd-nspawn.xml

```diff
@@ -747,31 +756,44 @@
       </varlistentry>
 
       <varlistentry>
-        <term><option>--private-users-chown</option></term>
-
-        <listitem><para>If specified, all files and directories in the container's directory tree will be
-        adjusted so that they are owned by the appropriate UIDs/GIDs selected for the container (see above).
-        This operation is potentially expensive, as it involves iterating through the full directory tree of
-        the container. Besides actual file ownership, file ACLs are adjusted as well.</para>
-
-        <para>This option is implied if <option>--private-users=pick</option> is used. This option has no effect if
-        user namespacing is not used.</para></listitem>
+        <term><option>--private-users-ownership=</option></term>
+
+        <listitem><para>Controls how to adjust the container image's UIDs and GIDs to match the UID/GID range
+        chosen with <option>--private-users=</option>, see above. Takes one of <literal>off</literal> (to
+        leave the image as is), <literal>chown</literal> (to recursively <function>chown()</function> the
+        container's directory tree as needed), <literal>map</literal> (in order to use transparent ID mapping
+        mounts) or <literal>auto</literal> for automatically using <literal>map</literal> where available and
+        <literal>chown</literal> where not.</para>
+
+        <para>If <literal>chown</literal> is selected, all files and directories in the container's directory
+        tree will be adjusted so that they are owned by the appropriate UIDs/GIDs selected for the container
+        (see above).  This operation is potentially expensive, as it involves iterating through the full
+        directory tree of the container. Besides actual file ownership, file ACLs are adjusted as
+        well.</para>
+
+        <para>Typically <literal>map</literal> is the best choice, since it transparently maps UIDs/GIDs in
+        memory as needed without modifying the image, and without requiring an expensive recursive adjustment
+        operation. However, it is not available for all file systems, currently.</para>
```

Comment:
should we document which ones - or point to kernel docs which will say which ones are supported?

Reply:
i figure this will change very quickly, not sure we should bother with this. Given that people use all kinds of kernels with all kinds of userspace I doubt we can comprehensively list what works and what doesn't in a way it's actually helpful

Reply:
That's fair - but is there some documentation on kernel.org of this feature, that we could point to? Or even a header on github.com/torvalds/linux would work. Ie, nothing specific, just a "go look there to figure it out"

Reply:
i don't think anyone compiles such a list. @brauner do you know?

Reply:
No, there's nothing. You can determine the supported filesystem by doing:
```
git grep FS_ALLOW_IDMAP

fs/ext4/super.c:        .fs_flags       = FS_REQUIRES_DEV | FS_ALLOW_IDMAP,
fs/fat/namei_msdos.c:   .fs_flags       = FS_REQUIRES_DEV | FS_ALLOW_IDMAP,
fs/fat/namei_vfat.c:    .fs_flags       = FS_REQUIRES_DEV | FS_ALLOW_IDMAP,
fs/namespace.c: if (!(m->mnt_sb->s_type->fs_flags & FS_ALLOW_IDMAP))
fs/xfs/xfs_super.c:     .fs_flags               = FS_REQUIRES_DEV | FS_ALLOW_IDMAP,
include/linux/fs.h:#define FS_ALLOW_IDMAP         32      /* FS has been updated to handle vfs idmappings. */
```

(We really need `fsinfo()` which should gain the ability to query a filesystem for the features it supports.)

Reply:
Fwiw, @tytso tried to expose that ext4 supports idmapped mounts by adding it to their feature file but that was shot down and would only work for a single filesystem so not optimal.

Reply:
Would it be possible to have a page on kernel.org docs in the future? It would really help with usability and user-friendliness. Something like fsverity does:

https://www.kernel.org/doc/html/latest/filesystems/fsverity.html#filesystem-support

Thanks!

Reply:
Sorry @bluca didn't ignore you yesterday I just missed your comment in the notification flood.

Yes, I think we want to do something like that. I also need to write a "porting guide" for filesystems.
And I need to make additional fs helper idmapped mount aware to e.g. port btrfs and overlayfs and others.

Reply:
@brauner i'd personally love to see support for this in btrfs + squashfs. btrfs is kinda important since it's what home defaults to since it's the only fs that can do online shrink + grow.

Reply:
> Sorry @bluca didn't ignore you yesterday I just missed your comment in the notification flood.
> 
> Yes, I think we want to do something like that. I also need to write a "porting guide" for filesystems.
> And I need to make additional fs helper idmapped mount aware to e.g. port btrfs and overlayfs and others.

Thanks! If/when that happens, we'd greatly appreciate a shout so that we can link to it here

Reply:
> @brauner i'd personally love to see support for this in btrfs + squashfs. btrfs is kinda important since it's what home defaults to since it's the only fs that can do online shrink + grow.

btrfs is definitely high-up my list too. It'll need some more vfs changes though.